crates/bevy_ecs/src/system/observer_system.rs

```diff
@@ -0,0 +1,114 @@
+use crate::{
+    prelude::{Bundle, Observer},
+    query::{QueryData, QueryFilter},
+    system::{System, SystemParam, SystemParamFunction, SystemParamItem},
+    world::DeferredWorld,
+};
+
+use super::{Commands, FunctionSystem, IntoSystem, Query, Res, ResMut, Resource};
+use bevy_utils::all_tuples;
+
+/// Implemented for systems that have an [`Observer`] as the first argument and only [`ObserverSystemParam`] implementors as parameters.
+pub trait ObserverSystem<E: 'static, B: Bundle>:
+    System<In = Observer<'static, E, B>, Out = ()> + Send + 'static
+{
+    /// Queues any deferred mutations to to be applied at the next [`apply_deferred`](crate::prelude::apply_deferred).
+    fn queue_deferred(&mut self, _world: DeferredWorld);
+}
+
+/// Implemented for [`SystemParam`] that can be used in [`Observer`] systems.
+/// # Safety
+/// Implementing types must ensure their implementation of [`SystemParam::get_param`] does not use `world` except as a [`DeferredWorld`]
+pub unsafe trait ObserverSystemParam: SystemParam {}
```

Comment:
I don't believe this trait is necessary. `SystemParams` are already forbidden from making structural changes to the world.

...as I'm scanning the docs, I don't think we've actually documented this anywhere. It is a critical safety requirement for all non-exclusive systems though.

Reply:
You've got a point, I only really need the additional method to queue commands over applying directly to &mut World.

Reply:
I believe part of my motivation was to mark params that wouldn't necessarily run `apply` but maybe that can be done with docs.

Reply:
We could also keep the trait and remove the unsafe. In a different PR we should add that invariant to `SystemParam`.